akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala

```diff
@@ -4,21 +4,36 @@
 
 package akka.persistence.journal
 
-import akka.persistence.{ PersistentRepr, PersistentEnvelope }
+import akka.persistence.{ Persistence, PersistentRepr, PersistentEnvelope }
 import akka.actor.Actor
 import scala.collection.immutable
 
 private[akka] trait WriteJournalBase {
   this: Actor ⇒
 
+  private lazy val eventAdapters = Persistence(context.system).adaptersFor(context.self)
+
   protected def preparePersistentBatch(rb: immutable.Seq[PersistentEnvelope]): immutable.Seq[PersistentRepr] =
-    rb.filter(persistentPrepareWrite).asInstanceOf[immutable.Seq[PersistentRepr]] // filter instead of flatMap to avoid Some allocations
+    rb.collect {
+      case p: PersistentRepr ⇒
+        p.prepareWrite()
+        adaptToJournal(p)
+    } // collect instead of flatMap to avoid Some allocations
+
+  /** INTERNAL API */
+  private[akka] final def adaptFromJournal(repr: PersistentRepr): PersistentRepr = {
+    val adapter = eventAdapters.get(repr.payload.getClass)
```

Comment:
Enjoy the midsommar! :wink: (over here it's raining) (no releases today, no worries)

One of the possible scenarios is (it also starts to play into the ideas needed for https://github.com/akka/akka/issues/17616):

"old app"
- `persist(OldAppEvent)`
- catch all "toJson" adapter
  - stored id is something, lets say 1001

We start the new version of the app:
- "we need to move away from OldAppEvents!"
  - they are all stored as json, that's what the adapter gets
  - the model has become complex and we want multiple adapters for different types that are "inside the json"
  - the adapter gets json to the fromJournal() - current impl it would for all of these have to be the 1001 adapter
- **new from here-on:** we add EventAdapters with `fromJournal: Json => NewAppEvent`
  - adapter knows based on something inside the json if it should emit events or not (we provide `EventSeq.empty`)
  - can emit "nothing"
  - allow 1:n which we did not yet address

How it could be done with current:
- was stored without adapter so adapterId is 0
- users are kind of stuck with that decision if we only allow picking the adapter by the adapter id, we could allow binding to the 0 adapterId (which in theory should be our "internal reserved id" for identity)...